src/pkg/packager/sources/new_test.go

```diff
@@ -5,49 +5,168 @@
+			expectedIdentify: "split",
+			expectedType:     &SplitTarballSource{},
+		},
+	}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
 
-var sources = []source{
-	{pkgSrc: "oci://ghcr.io/defenseunicorns/packages/init:1.0.0", srcType: "oci", source: ociS},
-	{pkgSrc: "sget://github.com/defenseunicorns/zarf-hello-world:x86", srcType: "sget", source: urlS},
-	{pkgSrc: "sget://defenseunicorns/zarf-hello-world:x86_64", srcType: "sget", source: urlS},
-	{pkgSrc: "https://github.com/defenseunicorns/zarf/releases/download/v1.0.0/zarf-init-amd64-v1.0.0.tar.zst", srcType: "https", source: urlS},
-	{pkgSrc: "http://github.com/defenseunicorns/zarf/releases/download/v1.0.0/zarf-init-amd64-v1.0.0.tar.zst", srcType: "http", source: urlS},
-	{pkgSrc: "zarf-init-amd64-v1.0.0.tar.zst", srcType: "tarball", source: tarballS},
-	{pkgSrc: "zarf-package-manifests-amd64-v1.0.0.tar", srcType: "tarball", source: tarballS},
-	{pkgSrc: "zarf-package-manifests-amd64-v1.0.0.tar.zst", srcType: "tarball", source: tarballS},
-	{pkgSrc: "some-dir/.part000", srcType: "split", source: splitS},
+			require.Equal(t, tt.expectedIdentify, Identify(tt.src))
+			ps, err := New(&types.ZarfPackageOptions{PackageSource: tt.src})
+			require.NoError(t, err)
+			require.IsType(t, tt.expectedType, ps)
+		})
+	}
 }
 
-func Test_identifySourceType(t *testing.T) {
-	for _, source := range sources {
-		actual := Identify(source.pkgSrc)
-		require.Equalf(t, source.srcType, actual, fmt.Sprintf("source: %s", source))
+func TestPackageSource(t *testing.T) {
+	t.Parallel()
+
+	b, err := os.ReadFile("./testdata/expected-pkg.json")
+	require.NoError(t, err)
+	expectedPkg := types.ZarfPackage{}
+	err = json.Unmarshal(b, &expectedPkg)
+	require.NoError(t, err)
+
+	ts := httptest.NewServer(http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		_, fp := path.Split(req.URL.Path)
+		f, err := os.Open(path.Join("testdata", fp))
+		if err != nil {
+			rw.WriteHeader(http.StatusNotFound)
+			return
+		}
+		defer f.Close()
+		io.Copy(rw, f)
+	}))
+
+	tests := []struct {
+		name   string
+		src    string
+		shasum string
+	}{
+		{
+			name: "local",
+			src:  "./testdata/zarf-package-wordpress-amd64-16.0.4.tar.zst",
+		},
+		{
+			name:   "http",
+			src:    fmt.Sprintf("%s/zarf-package-wordpress-amd64-16.0.4.tar.zst", ts.URL),
+			shasum: "7161a86e0f33e0273131ce60a205d7656472266edd6392f509de2879fc21c671",
+		},
 	}
-}
+	for _, tt := range tests {
+		tt := tt
+		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
+
+			opts := &types.ZarfPackageOptions{
+				PackageSource: tt.src,
+				Shasum:        tt.shasum,
+			}
+
+			ps, err := New(opts)
+			require.NoError(t, err)
+			packageDir := t.TempDir()
+			pkgLayout := layout.New(packageDir)
+			pkg, warnings, err := ps.LoadPackage(pkgLayout, filters.Empty(), false)
+			require.NoError(t, err)
+			require.Empty(t, warnings)
+			require.Equal(t, expectedPkg, pkg)
+
+			ps, err = New(opts)
```

Comment:
I dont really understand why the package source isn't idempotent. This is forcing me to create a new instance for every check. This should be explored in the future.

Reply:
At the time I wrote sources, idempotence was not a consideration I was thinking about.

https://github.com/defenseunicorns/zarf/blob/f2d360261da70e656b090aa2f3b9bf036cf1dc19/src/pkg/packager/sources/new.go#L22-L30

In retrospect, I should have, as it allows for muuch better testing + repeatability + fuzzing.

afaik, every `zarf package deploy | inspect | remove | pull | publish` still only call 1 of these load functions, with some being destructive.

Reply:
I will let the tests remain as they are, then we can revisit this in the future. The docs helped be understand what was happening. I just feel like this could cause problems in the future if used improperly.